Fix oncue data unavailable when genset disconnected

PeteRager opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
When the generator became disconnected from Oncue; entities would be reported in HASS as "--", now the entities will report as unavailable


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Make the entities return unavailable; when the genset is disconnected and the oncue cloud is reporting "--" back as the data values.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes ##80666
- This PR is related to issue: #80666
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

PeteRager committed:
Fix: Make oncue data unavailable when genset disconnected

PeteRager commented:
I need to perform one more round of system testing by disconnecting the ethernet from the Generator.  I will do that later today.

PeteRager commented:
Testing complete all sensors went unavailable and then came back.   One exception is Battery Voltage, where is the messages sent back they give a value of Zero.  This can be fixed by using the NetworkConnectionEstablished data as the availability trigger.

Here is what it looks like after I disconnected ethernet cable.  Seems to happen within a minute or two.

![oncue unavailable](https://user-images.githubusercontent.com/76050312/196996633-e607a6d7-7b3e-47f9-9c07-84c1c8446652.PNG)

Here is what is looks like when restored:

![oncue available](https://user-images.githubusercontent.com/76050312/196996991-9e354ba5-fb90-46fd-80e8-d338520e91b1.PNG)

bdraco commented on tests/components/oncue/__init__.py:
```diff
@@ -556,3 +820,28 @@ def _patcher():
+    @contextmanager
+    def _patcher():
+
+        with patch("homeassistant.components.oncue.Oncue.async_login",), patch(
```
```suggestion
        with patch("homeassistant.components.oncue.Oncue.async_login"), patch(
```

PeteRager committed:
Update tests/components/oncue/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

PeteRager committed:
Update tests/components/oncue/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

PeteRager committed:
Move availability logic into availability property

Fix bug with Battery Voltage.  Cloud reports 0.0 when generator is unavailable.  This is solved by using the connection status indicator in the availability function.

PeteRager commented on homeassistant/components/oncue/entity.py:
```diff
@@ -52,4 +52,6 @@ def _oncue_value(self) -> str:
         device: OncueDevice = self.coordinator.data[self._device_id]
         sensor: OncueSensor = device.sensors[self.entity_description.key]
+        if sensor.value == VALUE_UNAVAILABLE:
+            return STATE_UNAVAILABLE
```
I like it.  Changed it to use that property.  Also I fixed an issue with the availability of the battery voltage.  In my testing, the cloud reports a 0 for this when the generator is disconnected rather than "--" like all the other sensors.  Usually battery voltage is 13.4 and a zero value would mean a dead battery....  Anyways, added additional logic into the property to make data also unavailable if the cloud is saying the generator is disconnected.

bdraco commented on homeassistant/components/oncue/entity.py:
```diff
@@ -53,3 +53,19 @@ def _oncue_value(self) -> str:
+        """Return if entity is available."""
+        if self._oncue_value == VALUE_UNAVAILABLE:
+            return False
+        # If the cloud is reporting that the generator is not connected
+        # this also indicates the data is not available.
+        # The battery voltage sensor reports 0.0 rather than -- hence the purpose of this check.
+        # However, the binary sensor that tracks the connection should not go unavailable.
+        if self.entity_description.key != CONNECTION_ESTABLISHED_KEY:
+            device: OncueDevice = self.coordinator.data[self._device_id]
+            conn_established: OncueSensor = device.sensors[CONNECTION_ESTABLISHED_KEY]
+            if conn_established is not None and conn_established.value == "false":
+                return False
```
In a future PR it might be nice to subclass the coordinator and make this a property that gets set during updates so we don't have to do all the dict lookups for each entity when state is written.

PeteRager replied:
Agree

bdraco commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

bdraco commented on homeassistant/components/oncue/entity.py:
```diff
@@ -53,3 +53,19 @@ def _oncue_value(self) -> str:
+        if self.entity_description.key != CONNECTION_ESTABLISHED_KEY:
+            device: OncueDevice = self.coordinator.data[self._device_id]
+            conn_established: OncueSensor = device.sensors[CONNECTION_ESTABLISHED_KEY]
+            if conn_established is not None and conn_established.value == "false":
```
I pulled the network out and the value was `--`

```
                  {'displayname': 'Network Connection Established',
                   'displayvalue': '--',
                   'groupdisplayname': 'Network Status',
                   'groupid': '52',
                   'groupname': 'NetworkStatus',
                   'id': '549405279705',
                   'name': 'NetworkConnectionEstablished',
                   'parameterdefinitionid': '872',
                   'pdiid': '4018',
                   'recordstatus': 1,
                   'value': '--',
                   'version': '221'}],
```

PeteRager replied:
Thank you, that is a bug in the change.  I’ll fix tonite and retest pulling the cable.   I did not do that after the last change.

PeteRager replied:
Have that fixed.  Ran cable pulling test and it's working as it should now.

My guess is you saw Battery Voltage not go unavailable and the Connection States binary sensor go unavailable.

bdraco replied:
I'll give it a shot tomorrow when its light out again

bdraco replied:
I'm having trouble getting the generator back online after pulling it.  Looks like I'm going to have to reload the firmware as that usually fixes it.  🤦

bdraco replied:
Reloading the firmware got it back online, and there was a firmware update with a major version for it so hopefully its more reliable

PeteRager committed:
Fix for NetworkConnectionEstablished

This items return "--", not False when the generator goes offline.
Updated Availability Code
Updated unit test data with correct data
Ran system test with ethernet disconnect / reconnect
Add me to code owners

PeteRager committed:
Update CODEOWNERS

Add peterager to codeowners for oncue

PeteRager commented on homeassistant/components/oncue/entity.py:
```diff
@@ -53,3 +53,23 @@ def _oncue_value(self) -> str:
+    def available(self) -> bool:
+        """Return if entity is available."""
+        # The binary sensor that tracks the connection should not go unavailable.
+        if self.entity_description.key != CONNECTION_ESTABLISHED_KEY:
```
Yes, that would be cleaner.  I won’t be able to call super().available from there as this will wind up back here.   Concerns on that?

MartinHjelmare replied:
Right. Ok, then let's keep it like this.

bdraco commented:
Thanks @PeteRager
